app/api/groups/create_user_batch.go

```diff
@@ -0,0 +1,328 @@
+	createdUsers []loginmodule.CreateUsersResponseDataRow, subgroupsApprovals []subgroupApproval, user *database.User) []*resultRow {
+	result := make([]*resultRow, 0, len(subgroupsApprovals))
+
+	service.MustNotBeError(srv.Store.InTransaction(func(store *database.DataStore) error {
+		domainConfig := domain.ConfigFromContext(r.Context())
+
+		relationsToCreate := make([]map[string]interface{}, 0, 2*numberOfUsersToBeCreated)
+		usersToCreate := make([]map[string]interface{}, 0, numberOfUsersToBeCreated)
+		usersInSubgroup := 0
+		var currentResultRow *resultRow
+		currentSubgroupIndex := -1
+		for _, createdUser := range createdUsers {
+			createdUser := createdUser
+			if currentSubgroupIndex == -1 || usersInSubgroup == input.Subgroups[currentSubgroupIndex].Count {
+				currentSubgroupIndex++
+				currentResultRow = &resultRow{
+					GroupID: input.Subgroups[currentSubgroupIndex].GroupID,
+					Users:   make([]resultRowUser, 0, input.Subgroups[currentSubgroupIndex].Count),
+				}
+				result = append(result, currentResultRow)
+				usersInSubgroup = 0
+			}
+
+			var userGroupID int64
+			service.MustNotBeError(store.RetryOnDuplicatePrimaryKeyError(func(retryIDStore *database.DataStore) error {
+				userGroupID = retryIDStore.NewID()
+				return retryIDStore.Groups().InsertMap(map[string]interface{}{
+					"id":          userGroupID,
+					"name":        createdUser.Login,
+					"type":        "User",
+					"description": createdUser.Login,
+					"created_at":  database.Now(),
+					"is_open":     false,
+					"send_emails": false,
+				})
+			}))
+
+			var personalInfoApprovedAt, lockMembershipApprovedAt, watchApprovedAt interface{}
+			if subgroupsApprovals[currentSubgroupIndex].RequirePersonalInfoAccessApproval {
+				personalInfoApprovedAt = database.Now()
+			}
+			if subgroupsApprovals[currentSubgroupIndex].RequireLockMembershipApproval {
+				lockMembershipApprovedAt = database.Now()
+			}
+			if subgroupsApprovals[currentSubgroupIndex].RequireWatchApproval {
+				watchApprovedAt = database.Now()
+			}
+			relationsToCreate = append(relationsToCreate,
+				map[string]interface{}{
+					"parent_group_id":                domainConfig.RootSelfGroupID,
+					"child_group_id":                 userGroupID,
+					"personal_info_view_approved_at": nil, "lock_membership_approved_at": nil, "watch_approved_at": nil},
+				map[string]interface{}{
+					"parent_group_id":                input.Subgroups[currentSubgroupIndex].GroupID,
+					"child_group_id":                 userGroupID,
+					"personal_info_view_approved_at": personalInfoApprovedAt,
+					"lock_membership_approved_at":    lockMembershipApprovedAt,
+					"watch_approved_at":              watchApprovedAt},
+			)
+
+			usersToCreate = append(usersToCreate, map[string]interface{}{
+				"temp_user":     0,
+				"registered_at": database.Now(),
+				"group_id":      userGroupID,
+				"login_id":      createdUser.ID,
+				"login":         createdUser.Login,
+				"creator_id":    user.GroupID,
```

Comment:
default_language should probably be set as well

Reply:
fixed